src/algebra/jordan/triple.lean

```diff
@@ -0,0 +1,173 @@
+  ∀ (a b c : A),  T ⦃ a, b, c ⦄  = ⦃ T a, b, c⦄ + ⦃a, T' b, c⦄ + ⦃a, b, T c⦄
+
+namespace is_tp
+
+lemma radd {A : Type*} [has_tp A] [has_add A] [is_tp A] (a b c₁ c₂ : A) :
+  ⦃a, b, c₁ + c₂⦄ = ⦃a, b, c₁⦄ + ⦃a, b, c₂⦄ := by rw [comm, ladd, comm, comm c₂]
+
+variables {A : Type*} [has_tp A] [add_comm_group A] [is_tp A]
+
+lemma lzero (b c : A) : ⦃0, b, c⦄ = 0 :=
+(add_monoid_hom.mk' (λ (a : A), ⦃a, b, c⦄) (λ a₁ a₂, ladd a₁ a₂ b c )).map_zero
+
+lemma mzero (a c : A) : ⦃a, 0, c⦄ = 0 :=
+(add_monoid_hom.mk' (λ (b : A), ⦃a, b, c⦄) (λ b₁ b₂ , madd a b₁ b₂ c)).map_zero
+
+lemma rzero (a b : A) : ⦃a, b, 0⦄ = 0 :=
+by rw [comm, lzero]
+
+lemma lzsmul (a b c : A) (z : ℤ) : ⦃z•a, b, c⦄ = z•⦃a, b, c⦄ :=
+add_monoid_hom.map_zsmul ⟨λ (a : A), ⦃a, b, c⦄, lzero b c, λ _ _, ladd _ _ _ _⟩ _ _
+
+lemma mzsmul (a b c : A) (z : ℤ) : ⦃a, z•b, c⦄ = z•⦃a, b, c⦄ :=
+add_monoid_hom.map_zsmul ⟨λ (b : A), ⦃a, b, c⦄, mzero a c, λ _ _, madd _ _ _ _⟩ _ _
+
+lemma rzsmul (a b : A) (z : ℤ) (c : A) : ⦃a, b, z•c⦄ = z•⦃a, b, c⦄ :=
+  by rw [comm, lzsmul, comm]
+
+lemma lneg (a b c : A) : ⦃-a, b, c⦄ = -⦃a, b, c⦄ :=
+by rw [←sub_eq_zero, sub_neg_eq_add, ←ladd, neg_add_self, lzero]
+
+lemma mneg (a b c : A): ⦃a, -b, c⦄ = -⦃a, b, c⦄ :=
+by rw [←sub_eq_zero, sub_neg_eq_add, ←madd, neg_add_self, mzero]
+
+lemma rneg (a b c : A): ⦃a, b, -c⦄ = -⦃a, b, c⦄ := by rw [comm, lneg, comm]
+
+lemma lsub (a b c d : A) : ⦃a - d, b, c⦄ = ⦃a, b, c⦄ - ⦃d, b, c⦄ :=
+by rw [eq_sub_iff_add_eq, ← ladd, sub_add_cancel]
+
+lemma msub (a b c d : A) : ⦃a, b - c, d⦄ = ⦃a, b, d⦄ - ⦃a, c, d⦄ :=
+by rw [eq_sub_iff_add_eq, ← madd, sub_add_cancel]
+
+
+lemma rsub (a b c d : A) : ⦃a, b, c - d⦄ = ⦃a, b, c⦄ - ⦃a, b, d⦄ :=
+by rw [comm, lsub, comm, comm d]
+
+lemma lr_bilinear (a₁ a₂ b c₁ c₂ : A) : ⦃a₁ + a₂, b, c₁ + c₂⦄ =
+  ⦃a₁, b, c₁⦄ + ⦃a₁, b, c₂⦄ + ⦃a₂, b, c₁⦄ + ⦃a₂, b, c₂⦄ :=
+calc ⦃a₁ + a₂, b, c₁ + c₂⦄ = ⦃a₁, b, c₁ + c₂⦄ + ⦃a₂, b, c₁ + c₂⦄ : by rw ladd
+... = ⦃a₁, b, c₁⦄ + ⦃a₁, b, c₂⦄ + ⦃a₂, b, c₁ + c₂⦄ : by rw radd
+... = ⦃a₁, b, c₁⦄ + ⦃a₁, b, c₂⦄ + (⦃a₂, b, c₁⦄ + ⦃a₂, b, c₂⦄) : by rw radd
+... = ⦃a₁, b, c₁⦄ + ⦃a₁, b, c₂⦄ + ⦃a₂, b, c₁⦄ + ⦃a₂, b, c₂⦄ : by rw ← add_assoc
+
+lemma polar (a b c : A) : ⦃a + c, b, a + c⦄ = ⦃a, b, a⦄ + 2•⦃a, b, c⦄ + ⦃c, b, c⦄ :=
+calc ⦃a + c, b, a + c⦄ = ⦃a, b, a⦄ + ⦃a, b, c⦄ + ⦃c, b, a⦄ + ⦃c, b, c⦄ :
+  by rw lr_bilinear a c b a c
+... = ⦃a, b, a⦄ + ⦃a, b, c⦄ + ⦃a, b, c⦄ + ⦃c, b, c⦄ : by rw comm c b a
+... = ⦃a, b, a⦄ + (⦃a, b, c⦄ + ⦃a, b, c⦄)  + ⦃c, b, c⦄ : by rw ← add_assoc
+... = ⦃a, b, a⦄ + 2•⦃a, b, c⦄ + ⦃c, b, c⦄ : by rw two_nsmul
+
+/-- The triple product as an additive monoid homomorphism in each variable -/
+@[simps] def add_monoid_hom.tp : A →+ A →+ A →+ A :=
+{ to_fun := λ a,
+  { to_fun := λ b,
+    { to_fun := λ c, ⦃a, b, c⦄,
+      map_zero' := by rw rzero,
+      map_add' := λ _ _, by rw radd, },
+    map_zero' := add_monoid_hom.ext $ λ b, begin
+      simp only [add_monoid_hom.zero_apply, add_monoid_hom.coe_mk],
+      rw mzero,
```

Comment:
If you can replace these `rw`s with `exact`, then you can remove the `simp`s and save 12 lines here

Reply:
Do you mean:
```
   map_zero' := add_monoid_hom.ext $ λ b, begin
      --simp only [add_monoid_hom.zero_apply, add_monoid_hom.coe_mk],
       exact mzero,
    end,
```
?
This gives me:
```
expected type:
A: Type u_1
_inst_1: has_tp A
_inst_2: add_comm_group A
_inst_3: is_tp A
ab: A
⊢ ⇑{to_fun := λ (c : A), ⦃a,0,c⦄, map_zero' := _, map_add' := _} b = ⇑0 b
Messages (1)
triple.lean:126:13
invalid type ascription, term has type
  ∀ (a c : ?m_1), ⦃a,0,c⦄ = 0
but is expected to have type
  ⇑{to_fun := λ (c : A), ⦃a,0,c⦄, map_zero' := _, map_add' := _} b = ⇑0 b
state:
A : Type u_1,
_inst_1 : has_tp A,
_inst_2 : add_comm_group A,
_inst_3 : is_tp A,
a b : A
⊢ ⇑{to_fun := λ (c : A), ⦃a,0,c⦄, map_zero' := _, map_add' := _} b = ⇑0 b
```

Reply:
That error message is telling you that `mzero` is missing two arguments

Reply:
Fixed. Thanks!